### PR TITLE
08-daemon.py

### PR DESCRIPTION
**Pull request by Guy-Markman**

### CODE REVIEW

**Comment by Guy-Markman** [comment]
@alonbl
This is my Daemon until now

File: EX04/08-daemon.py [hunk 1]
```diff
@@ -1,10 +1,45 @@
...
 import os
+import sys
+import resource
+import time
+import signal
+import util
+
+
+def terminate_handler(signum, frame):
+    sys.exit()
```
Comment (alonbl) [inline EX04/08-daemon.py]: please do not exit in signal handler

File: EX04/08-daemon.py [hunk 2]
```diff
@@ -1,10 +1,45 @@
...
+    new_directory = os.open("/dev/null", os.O_RDWR)
+    for x in range(3):
+        os.dup2(new_directory, x)
+    os.chdir("/")
+    signal.signal(signal.SIGINT, terminate_handler)
+    signal.signal(signal.SIGTERM, terminate_handler)
+    signal.signal(signal.SIGHUP, signal.SIG_IGN)
+
+def proc_child_restarter (signum, frame):
+    proc_child()
```
Comment (alonbl) [inline EX04/08-daemon.py]: I do not  understand why this is required, please explain

File: EX04/08-daemon.py [hunk 3]
```diff
@@ -1,10 +1,45 @@
...
+
+def proc_child():
+    counter = 0
+    while True:
+        time.sleep(1)
```
Comment (alonbl) [inline EX04/08-daemon.py]: please use signals

File: EX04/08-daemon.py [hunk 4]
```diff
@@ -1,10 +1,45 @@
...
+        util.write_to_target(1, str(counter))
+        counter += 1
 
 
 def main():
-    
+    set_up()
```
Comment (alonbl) [inline EX04/08-daemon.py]: where is the log file?

Comment (alonbl) [inline EX04/08-daemon.py]: please tell me when you think your implementation meets all requirements

File: EX04/08-daemon.py [hunk 5]
```diff
@@ -7,37 +7,48 @@
...
+def terminate_handler(signum, frame):
+    exit_gracefully()
+
+
+def set_up():  # Return the FD of the log file
     new_directory = os.open("/dev/null", os.O_RDWR)
     for x in range(3):
         os.dup2(new_directory, x)
     os.chdir("/")
+    log_fd = os.open("log_file.txt", os.O_CREAT | os.O_WRONLY)
```
Comment (alonbl) [inline EX04/08-daemon.py]: we usually open logs in append mode

File: EX04/08-daemon.py [hunk 6]
```diff
@@ -7,37 +7,48 @@
...
     while True:
         time.sleep(1)
-        util.write_to_target(1, str(counter))
-        counter += 1
+        util.write_to_target(log_fd, str(sys.maxint - time_left()))
 
 
 def main():
-    set_up()
+    log_fd = set_up()
```
Comment (alonbl) [inline EX04/08-daemon.py]: where do you close this?

Comment (alonbl) [inline EX04/08-daemon.py]: I am unsure why you setup signals and such in parent

File: EX04/08-daemon.py [hunk 7]
```diff
@@ -7,37 +7,48 @@
...
     child = os.fork()
     if child == 0:
         signal.signal(signal.SIGUSR1, proc_child_restarter)
-        proc_child()
+        proc_child(log_fd)
```
Comment (alonbl) [inline EX04/08-daemon.py]: what if you have an error? where will we see it?

File: EX04/08-daemon.py [hunk 8]
```diff
@@ -7,37 +7,48 @@
...
     signal.signal(signal.SIGHUP, signal.SIG_IGN)
+    return log_fd
+
+
+def proc_child_restarter(signum, frame):
+    signal.setitimer(signal.ITIMER_REAL, sys.maxint)
+
+
+def time_left():
+    return int(signal.getitimer(signal.ITIMER_REAL))[0]
```
Comment (alonbl) [inline EX04/08-daemon.py]: hmmm... I expect somekind of counter without magic :)

File: EX04/08-daemon.py [hunk 9]
```diff
@@ -7,37 +7,48 @@
...
-        util.write_to_target(1, str(counter))
-        counter += 1
+        util.write_to_target(log_fd, str(sys.maxint - time_left()))
 
 
 def main():
-    set_up()
+    log_fd = set_up()
     child = os.fork()
     if child == 0:
```
Comment (alonbl) [inline EX04/08-daemon.py]: in daemon there is no concept of child as parent is actually gone, you can simplify your implementation greatly.

**Comment by Guy-Markman** [comment]
Should I fix things you wrote in outdated comments?

File: EX04/08-daemon.py (hunk 8 above)
Comment (Guy-Markman) [inline EX04/08-daemon.py]: How can I change a value using signal?
I gave you one solution I had. What's bad with this?

File: EX04/08-daemon.py (hunk 5 above)
Comment (Guy-Markman) [inline EX04/08-daemon.py]: Done

File: EX04/08-daemon.py (hunk 8 above)
Comment (Guy-Markman) [inline EX04/08-daemon.py]: I'm just upset because I think this is genius

File: EX04/08-daemon.py (hunk 6 above)
Comment (Guy-Markman) [inline EX04/08-daemon.py]: Changed in the last version

Comment (Guy-Markman) [inline EX04/08-daemon.py]: Changed

File: EX04/08-daemon.py [hunk 10]
```diff
@@ -1,10 +1,59 @@
...
 import os
+import sys
+import resource
+import time
+import signal
+import util
+
+
+def exit_gracefully():
+    sys.exit()
```
Comment (alonbl) [inline EX04/08-daemon.py]: once again, please do not exit within signal.

File: EX04/08-daemon.py [hunk 11]
```diff
@@ -1,10 +1,59 @@
...
+
+
+def terminate_handler(signum, frame):
+    exit_gracefully()
+
+
+def set_up():  # Return the FD of the log file
+    os.closerange(3, resource.RLIMIT_NOFILE)
+    new_directory = os.open("/dev/null", os.O_RDWR)
```
Comment (alonbl) [inline EX04/08-daemon.py]: what should you do with this file descriptor?

File: EX04/08-daemon.py [hunk 12]
```diff
@@ -1,10 +1,59 @@
...
+def time_left():
+    return int(signal.getitimer(signal.ITIMER_REAL))[0]
+
+
+def proc_child():
+    signal.signal(signal.SIGALRM, signal.SIG_IGN)
+    try:
+        with os.open("log_file.txt", os.O_CREAT | os.O_WRONLY | os.O_APPEND) as log_fd:
+            while True:
+                time.sleep(1)
```
Comment (alonbl) [inline EX04/08-daemon.py]: why 1?

File: EX04/08-daemon.py [hunk 13]
```diff
@@ -1,10 +1,59 @@
...
+
+
+def time_left():
+    return int(signal.getitimer(signal.ITIMER_REAL))[0]
+
+
+def proc_child():
+    signal.signal(signal.SIGALRM, signal.SIG_IGN)
+    try:
+        with os.open("log_file.txt", os.O_CREAT | os.O_WRONLY | os.O_APPEND) as log_fd:
```
Comment (alonbl) [inline EX04/08-daemon.py]: you do not need WRONLY if you specify APPEND

File: EX04/08-daemon.py [hunk 14]
```diff
@@ -1,10 +1,59 @@
...
+                util.write_to_target(log_fd, str(sys.maxsize - time_left()))
+    except Exception as e:
+        with os.open("log_file.txt", os.O_WRONLY | os.O_APPEND) as log_fd:
+            util.write_to_target(log_fd, "Error! %s" % e)
 
 
 def main():
-    
+    child = os.fork()
+    if child == 0:
```
Comment (alonbl) [inline EX04/08-daemon.py]: trivial/shortest branch first, please

**Comment by Guy-Markman** [comment]
I forgot to push the last commits for 3 days

File: EX04/08-daemon.py [hunk 15]
```diff
@@ -1,10 +1,63 @@
...
+    signal.signal(signal.SIGINT, set_none)
+    signal.signal(signal.SIGTERM, set_none)
+    signal.signal(signal.SIGHUP, signal.SIG_IGN)
+    signal.signal(signal.SIGUSR1, set_none)
+    signal.signal(signal.SIGALRM, set_none)
+
+
+def proc_child():
+    counter = 0
+    log_fd = os.open("log_file.txt", os.O_CREAT | os.O_APPEND)
```
Comment (alonbl) [inline EX04/08-daemon.py]: if you use O_CREAT you must specify file mode

File: EX04/08-daemon.py [hunk 16]
```diff
@@ -1,10 +1,63 @@
...
+    try:
+        while True:
+            if signal.getsignal(
```
Comment (alonbl) [inline EX04/08-daemon.py]: why do you get signals?

File: EX04/08-daemon.py [hunk 17]
```diff
@@ -1,10 +1,63 @@
...
+                    signal.SIGINT) is None or signal.getsignal(
+                    signal.SIGTERM) is None:
+                break
+            if signal.getsignal(signal.SIGALRM) is None:
+                util.write_to_target(log_fd, str(counter))
+                counter += 1
+                signal.signal(signal.SIGALRM, set_none)
```
Comment (alonbl) [inline EX04/08-daemon.py]: why do you mess with signals within loop?

File: EX04/08-daemon.py [hunk 18]
```diff
@@ -1,10 +1,63 @@
...
-import signals
 import os
+import sys
+import resource
+import signal
+import util
+
+
+def set_none(signum, frame):
+    signal.signal(signum, None)
```
Comment (alonbl) [inline EX04/08-daemon.py]: what does this exactly do?

Comment (Guy-Markman) [inline EX04/08-daemon.py]: set the handler of the signal to None.

File: EX04/08-daemon.py (hunk 17 above)
Comment (Guy-Markman) [inline EX04/08-daemon.py]: So I'll know if the signal happend already.

**Comment by Guy-Markman** [comment]
Wait. Can I use Global variables?

File: EX04/08-daemon.py [hunk 19]
```diff
@@ -6,48 +6,48 @@
...
 
 
-def set_up():  # Return the FD of the log file
+def set_up():
     os.closerange(3, resource.RLIMIT_NOFILE)
-    new_directory = os.open("/dev/null", os.O_RDWR)
-    for x in range(3):
-        os.dup2(new_directory, x)
-    os.close(new_directory)
+    new_directory = os.open("/dev/null", os.O_WRONLY)
```
Comment (alonbl) [inline EX04/08-daemon.py]: 1. you can use os.devnull which abstracts the name of null, as in POSIX it is /dev/null and in Windows it is nul:
2. something is terribly wrong in this open...

File: EX04/08-daemon.py [hunk 20]
```diff
@@ -6,48 +6,48 @@
...
     os.chdir("/")
-    signal.signal(signal.SIGINT, set_none)
-    signal.signal(signal.SIGTERM, set_none)
+    signal.signal(signal.SIGINT, set_ign)
+    signal.signal(signal.SIGTERM, set_ign)
     signal.signal(signal.SIGHUP, signal.SIG_IGN)
-    signal.signal(signal.SIGUSR1, set_none)
-    signal.signal(signal.SIGALRM, set_none)
+    signal.signal(signal.SIGUSR1, set_ign)
+    signal.signal(signal.SIGALRM, set_ign)
```
Comment (alonbl) [inline EX04/08-daemon.py]: I do not understand why you register the same handler to all signals.

File: EX04/08-daemon.py [hunk 21]
```diff
@@ -6,48 +6,48 @@
...
 
 
 def proc_child():
     counter = 0
-    log_fd = os.open("log_file.txt", os.O_CREAT | os.O_APPEND)
-    try:
+    log_fd = os.open("log_file.txt", os.O_CREAT | os.O_APPEND, 00777)
```
Comment (alonbl) [inline EX04/08-daemon.py]: why 00?
why 777?

File: EX04/08-daemon.py [hunk 22]
```diff
@@ -6,48 +6,48 @@
...
+                    signal.SIGTERM) is signal.SIG_IGN:
                 break
-            if signal.getsignal(signal.SIGALRM) is None:
+            if signal.getsignal(signal.SIGALRM) is signal.SIG_IGN:
                 util.write_to_target(log_fd, str(counter))
                 counter += 1
                 signal.signal(signal.SIGALRM, set_none)
                 signal.alarm(1)
-            if signal.getsignal(signal.SIGUSR1) is None:
+            if signal.getsignal(signal.SIGUSR1) is signal.SIG_IGN:
```
Comment (alonbl) [inline EX04/08-daemon.py]: what are you doing here exactly? I do not understand. why don't you use signal handler?

**Comment by Guy-Markman** [comment]
Sarit told us that you won't check it anymore.
Should I merge?

**Comment by alonbl** [comment]
Yes.

**Pull request merged**